File: vortex/backend/src/main/java/com/vortex/service/ChromaService.java
package com.vortex.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import java.util.Map;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;
import java.time.LocalDateTime;

@Service
@Slf4j
public class ChromaService {

    private final WebClient webClient;
    private final LiteLlmService liteLlmService;

    @Value("${chroma.url}")
    private String chromaUrl;

    @Value("${chroma.collection-name}")
    private String collectionName;

    public ChromaService(WebClient.Builder webClientBuilder, LiteLlmService liteLlmService) {
        this.webClient = webClientBuilder.build();
        this.liteLlmService = liteLlmService;
    }

    /**
     * Purges existing records for a file to prevent duplicates.
     */
    public void deleteByFilename(String filename) {
        try {
            String collectionId = getCollectionId();
            webClient.post()
                .uri(chromaUrl + "/api/v1/collections/" + collectionId + "/delete")
                .bodyValue(Map.of("where", Map.of("source", filename)))
                .retrieve()
                .toBodilessEntity()
                .block();
            log.info("Purged existing records for: {}", filename);
        } catch (Exception e) {
            log.error("Purge error for {}: {}", filename, e.getMessage());
        }
    }

    /**
     * Ingests a chunk with semantic metadata.
     */
    public void upsert(String text, List<Double> embedding, String filename, int index, String topic) {
        try {
            String id = filename + "_" + index;
            webClient.post()
                .uri(chromaUrl + "/api/v1/collections/" + getCollectionId() + "/upsert")
                .bodyValue(Map.of(
                    "ids", List.of(id),
                    "embeddings", List.of(embedding),
                    "metadatas", List.of(Map.of(
                        "source", filename,
                        "topic", topic,
                        "chunk", index,
                        "ingested_at", LocalDateTime.now().toString()
                    )),
                    "documents", List.of(text)
                ))
                .retrieve()
                .toBodilessEntity()
                .block();
        } catch (Exception e) {
            log.error("Upsert Error: {}", e.getMessage());
        }
    }

    /**
     * Semantic Search method designed for Agent Tool usage.
     */
    public String searchDocs(String query, String apiKey) {
        try {
            log.info("Agent triggered search: {}", query);
            List<Double> embedding = liteLlmService.getEmbedding(query, apiKey).block();
            if (embedding == null) return "Error: Embedding failed.";

            Map<String, Object> response = webClient.post()
                .uri(chromaUrl + "/api/v1/collections/" + getCollectionId() + "/query")
                .bodyValue(Map.of(
                    "query_embeddings", List.of(embedding),
                    "n_results", 4,
                    "include", List.of("documents", "metadatas")
                ))
                .retrieve()
                .bodyToMono(Map.class)
                .block();

            if (response == null || !response.containsKey("documents")) return "No relevant info found.";

            List<List<String>> documents = (List<List<String>>) response.get("documents");
            List<List<Map<String, Object>>> metadatas = (List<List<Map<String, Object>>>) response.get("metadatas");

            StringBuilder result = new StringBuilder();
            for (int i = 0; i < documents.get(0).size(); i++) {
                Map<String, Object> meta = metadatas.get(0).get(i);
                result.append(String.format("[File: %s | Topic: %s]\n%s\n---\n", 
                        meta.get("source"), meta.get("topic"), documents.get(0).get(i)));
            }
            return result.toString();

        } catch (Exception e) {
            return "Knowledge Retrieval Error: " + e.getMessage();
        }
    }

    /**
     * Legacy method for production pre-injection (now secondary to searchDocs).
     */
    public String queryKnowledge(String query, String apiKey) {
        return searchDocs(query, apiKey);
    }

    private String getCollectionId() {
        try {
            Map res = webClient.get().uri(chromaUrl + "/api/v1/collections/" + collectionName).retrieve().bodyToMono(Map.class).block();
            if (res != null && res.containsKey("id")) return (String) res.get("id");
        } catch (Exception e) {
            Map createRes = webClient.post().uri(chromaUrl + "/api/v1/collections").bodyValue(Map.of("name", collectionName)).retrieve().bodyToMono(Map.class).block();
            if (createRes != null) return (String) createRes.get("id");
        }
        return "";
    }
}

File: vortex/backend-py/services/chroma_service.py
import requests
import os
import uuid
import litellm
import logging
from datetime import datetime
from typing import List, Dict, Any, Optional

logger = logging.getLogger(__name__)

class ChromaService:
    def __init__(self):
        self.chroma_url = os.getenv("CHROMA_URL", "http://localhost:8001")
        self.collection_name = os.getenv("CHROMA_COLLECTION_NAME", "org_knowledge")
        self.proxy_base_url = os.getenv("LITELLM_PROXY_BASE_URL", "https://internal-llm-proxy.company.com")

    def _get_embedding(self, text: str, api_key: str) -> List[float]:
        response = litellm.embedding(
            model="text-embedding-3-small",
            input=[text],
            api_base=self.proxy_base_url,
            api_key=api_key
        )
        return response['data'][0]['embedding']

    def delete_by_filename(self, filename: str):
        collection_id = self._get_or_create_collection()
        if not collection_id: return
        payload = {"where": {"source": filename}}
        try:
            requests.post(f"{self.chroma_url}/api/v1/collections/{collection_id}/delete", json=payload)
        except Exception as e:
            logger.error(f"Failed to purge {filename}: {e}")

    def upsert(self, text: str, embedding: List[float], filename: str, chunk_index: int, topic: str):
        collection_id = self._get_or_create_collection()
        if not collection_id: return

        payload = {
            "ids": [f"{filename}_{chunk_index}"],
            "embeddings": [embedding],
            "metadatas": [{
                "source": filename,
                "topic": topic, # ENHANCED: Semantic Topic
                "ingested_at": datetime.now().isoformat(),
                "chunk": chunk_index
            }],
            "documents": [text]
        }
        requests.post(f"{self.chroma_url}/api/v1/collections/{collection_id}/upsert", json=payload)

    def search_docs(self, query: str, api_key: str, filename_filter: Optional[str] = None) -> str:
        try:
            embedding = self._get_embedding(query, api_key)
            collection_id = self._get_or_create_collection()
            
            payload = {
                "query_embeddings": [embedding],
                "n_results": 4,
                "include": ["documents", "metadatas"]
            }
            
            if filename_filter:
                payload["where"] = {"source": filename_filter}

            response = requests.post(f"{self.chroma_url}/api/v1/collections/{collection_id}/query", json=payload).json()
            
            if "documents" in response and response["documents"]:
                results = []
                for i in range(len(response["documents"][0])):
                    doc = response["documents"][0][i]
                    meta = response["metadatas"][0][i]
                    # Show the Intelligent Topic in the search results
                    results.append(f"[File: {meta['source']} | Topic: {meta['topic']}]\n{doc}")
                return "\n---\n".join(results)
            return "No matching documentation found."
        except Exception as e:
            return f"Search Error: {str(e)}"

    def _get_or_create_collection(self) -> str:
        try:
            res = requests.get(f"{self.chroma_url}/api/v1/collections/{self.collection_name}").json()
            if "id" in res: return res["id"]
        except: pass
        try:
            res = requests.post(f"{self.chroma_url}/api/v1/collections", json={"name": self.collection_name}).json()
            return res.get("id")
        except: return None